### DISCUSSION TITLE
RF-spoiling and Gradient Spoiling

### DISCUSSION THREAD

**Discussion by Aburas98** [JuliaHealth/KomaMRI.jl#529] [Q&A]
Hello! I'm designing an RF-spoiled GRE sequence. I noticed that the reconstructed phantom image does not follow our expectations, but there is no issue when scanning the same sequence on scanner. My question is, does KomaMRI support simulating RF-spoiling and Gradient spoiling effects?

This is the reconstructed brain phantom image:
![image](https://github.com/user-attachments/assets/8af6b0ec-ccd9-47b4-8125-fc7d4721bbf3)

This is the reconstructed image from data collected by scanner:
![image](https://github.com/user-attachments/assets/f3dd5c8f-e79b-4387-8cc1-be92654001b9)

This is the sequence file
[spiral_test_24rf_rfspoi.zip](https://github.com/user-attachments/files/18280415/spiral_test_24rf_rfspoi.zip)

**Comment by cncastillo** [comment]
Hi! KomaMRI does support RF and gradient spoiling, but you need to be careful and add additional isochromats in the directions you need as the default phantom is not very dense.

I am curious, is the in-vivo image acquired with exactly the same sequence? The FOVs does not seem to match.

For RF spoiling, as everything is happening in the z-direction (slice-selection), you can increase the number of spins in that direction with:

```julia
obj_ui[] = brain_phantom3D(; us=[1, 1, 2], start_end=[175, 185]) # This modifies the object in the UI
``` 
If you do gradient spoiling, you may need to increase `us` for x and y.

As this spiral has 1600 points per TR, this may go beyond your GPU memory (#519), so you may need to change your simulation parameters:

```julia
sim_params = KomaMRICore.default_sim_params()
sim_params["Nblocks"] = 4_000 # You can play with this, more blocks is slower but lowers memory usage
```
To modify the simulation parameters in the UI:
```julia
KomaUI(sim = sim_params) # You can also change the recon parameters with `rec = `
```
With this setup I get:
![newplot](https://github.com/user-attachments/assets/3c877d25-7701-429c-bc83-d1648f2ddc5d)
This looks much more reasonable.

**Reply by Aburas98** [reply]
Thank you very much @cncastillo for the detailed answer!

Yes the sequence is exactly the same, FOV might be different because I increased it during reconstruction for both simulation and in-vivo image to see what kind of aliasing happens outside FOV. The sequence is designed for a 200 mm^2 FOV and 1mm isotropic resolution.

For RF-spoiling, I was trying to use it on the 2D phantom, not the 3D one, as I did not consider that RF-spoiling will have effect only in the ss-direction. I will try to change the phantom settings and see if there is improvement.

Why would you think that this image is more reasonable? Because my goal was to get an artifact-free image after using RF-spoiling, but this image still contains these swirling-patterns which theoretically shouldn't be there (at least to the best of my understanding)

**Reply by cncastillo** [reply]
I say more reasonable because there's no apparent simulation artifact in the reconstructed image. For me, the swirling looks like an under-sampling artifact, but using more spins in z and reconstruct the image externally is worth trying.

**Reply by Aburas98** [reply]
Oh I understand! This sequence shouldn't have any under-sampling, but it may be due to me setting incorrect simulation parameters.

I was using a 2D phantom and with small number of samples. I did not mind the small number of spins when using RF-spoiling, since I thought RF-spoiling should have effect on individual spins and the interaction between spins shouldn't be affected by RF-spoiling. **Using a 3D phantom and up-sampling as you suggested does indeed improve the simulation results significantly.** 
However, I still get a non-ideal simulation results when using the brain phantom. When I change the T2 of the phantom so that the signal decays before the next TR, these artifacts disappear (Of course the contrast will change as well when changing T2). 

I used two sequences: 
1- a spoiled GRE spiral sequence with 24 interleaves: "spiral_training_test_TRis2TE.seq"
2- a SSFP spiral sequence (same as the other sequence but no RF-spoiling): "spiral_training_test_TRis2TE_ssfp.seq"

TE = 10ms, TR = 20ms, FA = 15 for both sequences.

A- When using the default phantom parameters (simulation script shared at the end), I get the following images:

For the SSFP sequence:
![img_v3_02i6_40fd1a0f-224f-4b05-93f7-9d38b14faf8g](https://github.com/user-attachments/assets/fb619e5c-947a-4d55-b0e2-6b6cbe661f3e)

For the spoiled GRE sequence:
![img_v3_02i6_d19c6ad6-0c48-421a-8b0c-901a94b3c46g](https://github.com/user-attachments/assets/8b400d5e-b4bf-46c3-a7b2-d352078f2161)

B- When I change the phantom T2 to be 10% of its original value:

For the SSFP sequence:
![img_v3_02i6_ac05de26-bc9f-4ed1-8942-95d02d64ad5g](https://github.com/user-attachments/assets/55e9ecf1-c1c0-4603-b695-7e0ec2da82ae)

For the spoiled GRE sequence:
![img_v3_02i6_4013031e-20f5-4715-8f6a-1f8d8cff0c7g](https://github.com/user-attachments/assets/7ec257be-0ba0-4798-96f1-9f6c9bd557de)

-------------------------------------------------------------
I understand that in B case, Mxy is already decayed at the end of each TR. And this does not reflect real-life scenarios.

I uploaded here the sequence files with the simulation script written on a Pluto Notebook:
[KomaMRI_simulation.zip](https://github.com/user-attachments/files/18298481/KomaMRI_simulation.zip)


Right now, I will try to reconstruct the image externally as you suggested!



Thank you very much for the help and creating this amazing tool!

**Reply by cncastillo** [reply]
Wow! the results are looking pretty good! :smile: 

>  However, I still get a non-ideal simulation results when using the brain phantom. When I change the T2 of the phantom so that the signal decays before the next TR, these artifacts disappear (Of course the contrast will change as well when changing T2).

 I think @gsahonero has encountered something similar. @gsahonero  could check if you can help?

**Reply by gsahonero** [reply]
Hi everyone! 
Sorry for the delay. I took some days off. 

Yes, @cncastillo ! It is something similar. I haven't tested the code you shared @Aburas98 , but from what I experienced you can reduce the resolution and the FOV, and use the right spoiling strategy (without changing the T2). Also, while spoiling won't be effective in removing the artifacts, a wrongly executed spoiling generates distorted images - as expected. RF-spoiling is standard, but gradient spoiling can be implemented using different approaches. To check the effect of the spoiling, you could check the final `Mxy` using something like this: https://github.com/gsahonero/KomaMRI.jl/blob/transverse_magnetization_observation/KomaMRIPlots/src/ui/DisplayFunctions.jl#L1567 and this: https://github.com/gsahonero/KomaMRI.jl/blob/transverse_magnetization_observation/KomaMRICore/src/simulation/SimulatorCore.jl#L412. In any case, I'll check your code soon.

On the other hand, I'm curious: is the scanner using the pulseq file or are you using the vendor-available sequence with the TE/TR/FA you mentioned?

**Reply by Aburas98** [reply]
Hi @gsahonero! Thank you very much for the suggestions! I will try it and see how the RF-spoiling works.

The TR TE FA values are provided for the simulated two sequences I provided in my last reply. As for the sequence I showed in my first question "[spiral_test_24rf_rfspoi.zip]", I used the exact same .seq file for simulation and scanner.

**Comment by markymarkymark** [comment]
Joining in on this thread, as I also want to simulate RF and gradient spoiling. 

I don't follow how more spins are needed for RF spoiling (which is just incrementing the RF and ADC phase) to NOT create incorrect ghosting artifacts. 

To wit, even with TR >> T2*, a GRE simulation in KomaUI which applies RF spoiling produces an incorrectly ghosted image. Turning **off** RF spoiling (i.e. RF phase = ADC phase = constant for all TR) results in a clean image.

**Reply by cncastillo** [reply]
Hi! Sorry for the delay! I totally missed this comment. Feel free to tag me if I don't respond. 

Can you share the Pulseq sequence you used to be able to help with this problem? The number of isochromats in the z-direction for RF spoiling depends on how big the spoiler is. For gradient spoiling, the direction of the spoiler is important to know which directions might need more isochromats.
